### ISSUE TITLE
[Feature][Spec][XPU] Honor rejection sampling in EAGLE chain verification

### ISSUE THREAD

**Issue by adi271213** [sgl-project/sglang#41070]
### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Motivation

EAGLE verification on XPU takes the greedy argmax path even when the batch is non-greedy and --speculative-use-rejection-sampling is enabled. For a configuration that passes the startup checks, the startup hook reports rejection sampling as enabled, but the XPU verification gate still prevents that sampler from being reached.

I would like to enable the existing chain-rejection path on XPU, subject to hardware validation and coordination with related work.

Current behavior

Source snapshot: 32290dda2cea4bb95274b3d08e43d4dad74e9676.

In python/sglang/srt/speculative/eagle_utils.py, _verify_uses_greedy() returns:

return is_all_greedy or is_cpu or is_xpu or (
    is_hip and not use_rejection_sampling
)

The unconditional is_xpu term selects the branch containing torch.argmax(next_token_logits, dim=-1), including when is_all_greedy=False and use_rejection_sampling=True.

Meanwhile, speculative_hook.py logs:

Rejection sampling is enabled for speculative decoding (speculative_use_rejection_sampling=True).

The issue is therefore a mismatch between the enabled configuration and the verification path actually selected.

This is a source-level finding. I have not yet reproduced the behavior or validated an alternative on XPU hardware. The ignored flag was also noted as a side issue in #35144.

Proposed scope

Allow non-greedy XPU batches with rejection sampling explicitly enabled to reach chain_speculative_sampling_triton.

The existing startup checks and sampling-branch validation would remain unchanged:

The resolved algorithm must be EAGLE or EAGLE3, including NEXTN configurations that resolve to EAGLE.

--speculative-eagle-topk must be 1.

Both acceptance thresholds must be 1.0, and deterministic inference must be disabled.

Draft proposal probabilities must be present and use the target vocabulary.

These are inherited requirements of the existing rejection-sampling path, not additional XPU-specific restrictions. The speculative branching factor is separate from request-level sampling top_k.

Automatic enablement and the separate target-only stochastic tree-verification path are outside this proposal. Greedy requests and other backends should retain their existing behavior.

Related implementation work

#37134 provides a reference for enabling the Triton chain-rejection path on ROCm.

intel/llm-scaler#719, merged September 17, 2026, implements out-of-tree target-only stochastic XPU verification using SYCL/PyTorch operations and a patch to eagle_info_v2.py. It addresses the broader XPU sampling limitation, but is distinct from enabling the existing Triton chain sampler with actual draft proposal probabilities.

#34140 is also related to the separate target-only tree path on ROCm. I would like to coordinate with these efforts rather than duplicate planned upstream work.

Implementation considerations

Once XPU reaches the rejection-sampling branch, the existing renormalization dispatch already imports top_k_renorm_prob and top_p_renorm_prob from sgl_kernel.

The XPU dependency file pins sglang-kernel-xpu 0.2.0. I would first verify that this pinned wheel exposes working versions of the required operations and validate their cutoff and tie semantics against the intended sampling behavior. Related XPU sampler integration: #38510.

The Triton/PyTorch renormalization fallback is an alternative, not the existing XPU route. Using it would require explicit dispatch changes and an XPU-compatible _prepare_probs guard.

The sampling branch also broadcasts verification results across tensor-parallel ranks. TP > 1 therefore needs explicit validation in light of #35144/#35195; enabling the gate alone would not establish multi-XPU correctness.

Validation

Validation should include:

Gate-routing and configuration regression tests, including unchanged greedy behavior and the existing unsupported-configuration checks.

Actual XPU chain-sampler and renormalization tests against independent references, covering acceptance, rejection, all-drafts-accepted sampling, and top-k/top-p boundary and tie cases.

End-to-end non-greedy and mixed greedy/non-greedy batches, with distributional checks rather than output diversity or task accuracy alone.

Graph capture enabled/disabled and TP=1/TP>1 where supported, preserving and testing the existing draft_probs copy after graph replay so the proposal probabilities remain intact until verification.

Maintainer input

Is there an upstreaming plan for intel/llm-scaler#719 or another XPU verification effort that this should coordinate with? Would a focused in-tree chain-rejection enablement PR be useful alongside the target-only work?

I do not currently have guaranteed XPU hardware access. Before committing to an implementation, I would appreciate guidance on the preferred approach and whether hardware validation could be coordinated with an XPU contributor.
